src/midi.c

```diff
@@ -0,0 +1,979 @@
+                    ((uint8_t)mregs[sel].lsr_pe << 2) |
+                    ((uint8_t)mregs[sel].lsr_oe << 1) |
+                    ((uint8_t)(mregs[sel].ifsz > 0)));
+            if (!debugOn) {
+                mregs[sel].lsr_oe = false;
+                mregs[sel].lsr_pe = false;
+                mregs[sel].lsr_fe = false;
+                mregs[sel].lsr_bi = false;
+            }
+            return ret;
+            break;
+        }
+        case 0x6:
+            return (((uint8_t)mregs[sel].msr_dcd << 7) |
+                    ((uint8_t)mregs[sel].msr_ri << 6) |
+                    ((uint8_t)mregs[sel].msr_dsr << 5) |
+                    ((uint8_t)mregs[sel].msr_cts << 4) |
+                    ((uint8_t)mregs[sel].msr_ddcd << 3) |
+                    ((uint8_t)mregs[sel].msr_teri << 2) |
+                    ((uint8_t)mregs[sel].msr_ddsr << 1) |
+                    ((uint8_t)mregs[sel].msr_dcts));
+            break;
+        case 0x7:
+            return mregs[sel].scratch;
+            break;
+    }
+    return 0x00;
+}
+
+void midi_serial_calculate_clk(uint8_t sel)
+{
+    double uart_clks_per_cpu = (MIDI_UART_OSC_RATE_MHZ / MIDI_UART_PRIMARY_DIVIDER) / MHZ;
+    if (mregs[sel].dl > 0) {
+        uart_clks_per_cpu /= mregs[sel].dl;
+        mregs[sel].clockdec = uart_clks_per_cpu * 0x1000000L; // convert to 8.24 fixed point
+    } else {
+        mregs[sel].clockdec = 0;
+    }
+}
+
+void midi_serial_write(uint8_t reg, uint8_t val)
+{
+    time_t now = time(NULL);
+
+    //printf("midi_serial_write %d %d\n", reg, val);
+    uint8_t sel = (reg & 8) >> 3;
+    switch (reg & 7) {
+        case 0x0:
+            if (mregs[sel].lcr_dlab) {
+                mregs[sel].dll = val;
+                midi_serial_calculate_clk(sel);
+            } else {
+                if (mregs[sel].lcr_word_length_bits != 8 || mregs[sel].lcr_stb || mregs[sel].lcr_pen) {
+                    if (mregs[sel].last_warning + 60 < now) {
+                        unsigned char par = 'N';
+                        if (mregs[sel].lcr_pen) {
+                            switch ((uint8_t)(mregs[sel].lcr_eps << 1) | (uint8_t)mregs[sel].lcr_stick) {
+                                case 0:
+                                    par = 'O';
+                                    break;
+                                case 1:
+                                    par = 'M';
+                                    break;
+                                case 2:
+                                    par = 'E';
+                                    break;
+                                case 3:
+                                    par = 'S';
+                                    break;
+                            }
+                        }
+                        fprintf(stderr, "Serial MIDI: Warning: improper LCR %d%c%d for UART %d, must be set to 8N1.\n", mregs[sel].lcr_word_length_bits, par, 1+(uint8_t)mregs[sel].lcr_stb, sel);
```

Comment:
Who is this warning geared towards? People developing MIDI software or regular users that look at their program output? If I read that message on a terminal, I'd have no idea what LCR is, or whom I should report that to.

Reply:
This is meant for 6502 programmers trying to write software to talk to the MIDI card, and it means they got something wrong.  I figure this is better than no message and no MIDI sound (which is what they'd get on hardware) or it works despite the wrong settings (which would lull the programmer into the false notion that they got it right)